nx/lib/nx.ex

```diff
@@ -6658,6 +6658,251 @@ defmodule Nx do
+
+  ### Caveats
+
+  For big values of p, f64 rounding errors come into play
+  """
+
+  def norm(t_raw, input_opts \\ []) when is_list(input_opts) do
+    t = tensor!(t_raw)
+    assert_keys!(input_opts, [:ord, :axes, :keep_axes])
+
+    default_axes_opts = [axes: nil, keep_axes: false, ord: nil]
+    opts = Keyword.merge(default_axes_opts, input_opts)
+
+    case shape(t) do
+      {_, _} ->
+        :ok
+
+      {_} ->
+        :ok
+
+      s ->
+        raise "expected 1-D or 2-D tensor. Received a tensor with shape #{inspect(s)} instead."
+    end
+
+    rank = rank(t)
+    ord = get_norm_ord!(opts, rank)
+
+    axes_opts = Keyword.take(opts, [:axes, :keep_axes])
+
+    do_p_norm(t, ord, rank, axes_opts)
+  end
+
+  defp get_norm_ord!(opts, dims) do
+    ord = Keyword.get(opts, :ord)
+
+    case {ord, dims} do
+      {nil, _} ->
+        2
+
+      {:frobenius, 2} ->
+        2
+
+      {:frobenius, _} ->
+        raise "expected a 2-D tensor. Got a #{dims}-D tensor."
+
+      {:nuclear, _} ->
+        raise "nuclear norm not implemented yet."
+
+      {ord, 2} when ord in [:inf, :neg_inf] ->
+        ord
+
+      {ord, 1} when ord in [:inf, :neg_inf] ->
+        ord
+
+      {0, 2} ->
+        raise "ord #{inspect(ord)} not implemented for 2-D tensor."
+
+      {0, 1} ->
+        0
+
+      {ord, _} ->
+        ord
+    end
+  end
+
+  defp do_p_norm(t, 0, 1, opts) do
+    t
+    |> equal(0)
+    |> logical_not()
+    |> sum(opts)
+  end
+
+  defp do_p_norm(_t, 0, 2, _opts) do
+    raise "expected 1-D tensor for ord: 0. Got: 2-D tensor."
+  end
+
+  defp do_p_norm(t, ord, 2, axes_opts) when ord in [:inf, :neg_inf] do
+    opts = Keyword.merge(axes_opts, axes: [1])
+
+    function = if ord == :inf, do: &reduce_max/1, else: &reduce_min/1
+
+    t
+    |> Nx.abs()
+    |> sum(opts)
+    |> function.()
+  end
+
+  defp do_p_norm(t, ord, 1, axes_opts) when ord in [:inf, :neg_inf] do
+    function = if ord == :inf, do: &reduce_max/2, else: &reduce_min/2
+
+    t
+    |> Nx.abs()
+    |> function.(axes_opts)
+  end
+
+  defp do_p_norm(t, ord, 2, opts) when ord in [1, -1] do
+    function =
+      if ord == 1 do
+        &reduce_max/2
+      else
+        &reduce_min/2
+      end
+
+    t
+    |> Nx.abs()
+    |> sum(axes: [0])
+    |> function.(Keyword.take(opts, [:keep_axes]))
+  end
+
+  defp do_p_norm(_t, ord, 2, _opts) when ord not in -1..2 do
+    raise "invalid ord for 2-D tensor. Got: #{inspect(ord)}"
+  end
+
+  defp do_p_norm(%{type: type} = t, ord, _, opts) do
+    inv_ord = divide(1, ord)
+
+    # This coefficient is introduced for better numerical stability
+    # The idea is that by dividing the tensor by it, large values of
+    # tensor entries and large values of p are reduced, which in turn
+    # avoids numerical overflow.
+    {:ok, numerical_stability_coefficient} =
+      type
+      |> Nx.Type.max_value_binary()
+      |> Nx.from_binary(type)
+      |> Nx.Tensor.fetch(0)
```

Comment:
You don't need the fetch here, the max value scalar will work fine in subsequent operations!

Reply:
The intention here was to avoid having a [1]-shaped Tensor  instead of a pure scalar (altough they are almost the same) in some of the test results.

Reply:
I've simplified this __a lot__, so we can disregard my comment :)